File: Objective-C Version/KYPingTransitionDemo/KYPingTransition/SecondViewController.m
//
//  SecondViewController.m
//  KYPingTransition
//
//  Created by Kitten Yang on 1/31/15.
//  Copyright (c) 2015 Kitten Yang. All rights reserved.
//

#import "ViewController.h"
#import "SecondViewController.h"
#import "PingInvertTransition.h"

@interface SecondViewController ()

@end

@implementation SecondViewController{
    UIPercentDrivenInteractiveTransition *percentTransition;
}
-(void)viewWillAppear:(BOOL)animated{
    self.navigationController.delegate = self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    UIScreenEdgePanGestureRecognizer *edgeGes = [[UIScreenEdgePanGestureRecognizer alloc]initWithTarget:self action:@selector(edgePan:)];
    edgeGes.edges = UIRectEdgeLeft;
    [self.view addGestureRecognizer:edgeGes];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}


- (id <UIViewControllerInteractiveTransitioning>)navigationController:(UINavigationController *)navigationController
                          interactionControllerForAnimationController:(id <UIViewControllerAnimatedTransitioning>) animationController {
    return percentTransition;
}


-(void)edgePan:(UIPanGestureRecognizer *)recognizer{
    CGFloat per = [recognizer translationInView:self.view].x / (self.view.bounds.size.width);
    per = MIN(1.0,(MAX(0.0, per)));
    
    if (recognizer.state == UIGestureRecognizerStateBegan) {
        percentTransition = [[UIPercentDrivenInteractiveTransition alloc]init];
        [self.navigationController popViewControllerAnimated:YES];
    }else if (recognizer.state == UIGestureRecognizerStateChanged){
        [percentTransition updateInteractiveTransition:per];
    }else if (recognizer.state == UIGestureRecognizerStateEnded || recognizer.state == UIGestureRecognizerStateCancelled){
        if (per > 0.3) {
            [percentTransition finishInteractiveTransition];
        }else{
            [percentTransition cancelInteractiveTransition];
        }
        percentTransition = nil;
    }
}


- (id <UIViewControllerAnimatedTransitioning>)navigationController:(UINavigationController *)navigationController
                                   animationControllerForOperation:(UINavigationControllerOperation)operation
                                                fromViewController:(UIViewController *)fromVC
                                                  toViewController:(UIViewController *)toVC{
    if (operation == UINavigationControllerOperationPop) {
        PingInvertTransition *pingInvert = [PingInvertTransition new];
        return pingInvert;
    }else{
        return nil;
    }
}

- (IBAction)popClicked:(id)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

@end

File: Swift Version/KYPingTransition-Swift/KYPingTransition-Swift/SecondViewController.swift
//
//  SecondViewController.swift
//  KYPingTransition-Swift
//
//  Created by Kitten Yang on 1/8/16.
//  Copyright © 2016 Kitten Yang. All rights reserved.
//

import UIKit

class SecondViewController: UIViewController {

    private var percentTransition: UIPercentDrivenInteractiveTransition?
    
    // MARK : Lift Circle
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let edgeGes = UIScreenEdgePanGestureRecognizer(target: self, action: "edgePan:")
        edgeGes.edges = .Left
        view.addGestureRecognizer(edgeGes)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.delegate = self
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK : Gesture Method
    
    @objc private func edgePan(recognizer: UIPanGestureRecognizer) {

        var per = recognizer.translationInView(view).x / view.bounds.size.width
        per = min(1.0, max(0.0, per))
        
        if recognizer.state == .Began {
            percentTransition = UIPercentDrivenInteractiveTransition()
            navigationController?.popViewControllerAnimated(true)
        } else if recognizer.state == .Changed {
            percentTransition?.updateInteractiveTransition(per)
        } else if recognizer.state == .Cancelled || recognizer.state == .Ended {
            if per > 0.3 {
                percentTransition?.finishInteractiveTransition()
            } else {
                percentTransition?.cancelInteractiveTransition()
            }
            percentTransition = nil
        }
    }
    
    
    @IBAction func didTapBackButton(sender: AnyObject) {
        navigationController?.popViewControllerAnimated(true)
    }
}

extension SecondViewController: UINavigationControllerDelegate {
    func navigationController(navigationController: UINavigationController, interactionControllerForAnimationController animationController: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return percentTransition
    }

    func navigationController(navigationController: UINavigationController, animationControllerForOperation operation: UINavigationControllerOperation, fromViewController fromVC: UIViewController, toViewController toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        if operation == .Pop {
            let pingInvert = PingInvertTransition()
            return pingInvert
        } else {
            return nil
        }
    }
}


